### PR TITLE
TCP Tunneling: Optionally propagate response trailers to downstream info

### PR DESCRIPTION
**Pull request by IssaAbuKalbein**
Commit Message: Add support to save response trailers from CONNECT tunnels in tcp_proxy
Additional Description: -
Risk Level: low (opt-in)
Testing: unit, integration
Docs Changes: none
Release Notes: yes

### CODE REVIEW

**Comment by KBaichoo** [comment]
/assign @ggreenway 

/wait

For CI fixes

**Comment by alyssawilk** [comment]
Over to @kyessenov for first pass

File: source/common/tcp_proxy/tcp_proxy.cc [hunk 1]
```diff
@@ -570,13 +570,29 @@ const std::string& TunnelResponseHeaders::key() {
   CONSTRUCT_ON_FIRST_USE(std::string, "envoy.tcp_proxy.propagate_response_headers");
 }
 
+ProtobufTypes::MessagePtr TunnelResponseTrailers::serializeAsProto() const {
```
Comment (kyessenov) [inline source/common/tcp_proxy/tcp_proxy.cc:573]: Can you share the implementation with header version via a common class?

File: source/common/tcp_proxy/tcp_proxy.h [hunk 2]
```diff
@@ -113,21 +113,43 @@ using RouteConstSharedPtr = std::shared_ptr<const Route>;
...
  * Response headers for the tunneling connections.
  */
-class TunnelResponseHeaders : public StreamInfo::FilterState::Object {
+class TunnelResponseHeaders : public TunnelResponseHeadersOrTrailers {
 public:
   TunnelResponseHeaders(Http::ResponseHeaderMapPtr&& response_headers)
       : response_headers_(std::move(response_headers)) {}
-  const Http::ResponseHeaderMap& value() const { return *response_headers_; }
-  ProtobufTypes::MessagePtr serializeAsProto() const override;
+  const Http::ResponseHeaderMap& value() const override { return *response_headers_; }
```
[the remark this answers is not in the document]
Comment (IssaAbuKalbein) [reply source/common/tcp_proxy/tcp_proxy.h:132]: Why this change, please?

**Comment by IssaAbuKalbein** [comment]
@yanavlasov @alyssawilk @kyessenov Can we merge this PR, please?

**Review by kyessenov — DISMISSED** [review]
LGTM
@yanavlasov can you respond to the co-variant return type suggestion?

**Comment by kyessenov** [comment]
/wait
for Yan to close the comment.

File: source/common/tcp_proxy/tcp_proxy.h [hunk 3]
```diff
@@ -113,21 +113,43 @@ using RouteConstSharedPtr = std::shared_ptr<const Route>;
...
 };
 
+/**
+ * Response trailers for the tunneling connections.
+ */
+class TunnelResponseTrailers : public TunnelResponseHeadersOrTrailers {
+public:
+  TunnelResponseTrailers(Http::ResponseTrailerMapPtr&& response_trailers)
+      : response_trailers_(std::move(response_trailers)) {}
+  const Http::ResponseTrailerMap& value() const override { return *response_trailers_; }
```
[the remark this answers is not in the document]
Comment (IssaAbuKalbein) [reply source/common/tcp_proxy/tcp_proxy.h:146]: @yanavlasov, Why this change, please?

**Comment by ohadvano** [comment]
@yanavlasov can you please resolve the open comment? We could use this feature in the upcoming release.

File: source/common/tcp_proxy/tcp_proxy.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (IssaAbuKalbein) [reply source/common/tcp_proxy/tcp_proxy.h:132]: @yanavlasov, What you suggest is a breaking change for those are using "value()" with the return type "ResponseHeaderMap&".

**Comment by IssaAbuKalbein** [comment]
I had pushed a commit not signed off, and tried to fix it but made it did not work to me after many tries :(
Can you, please, move to review that is one: https://github.com/envoyproxy/envoy/pull/25023

**Pull request closed without merging**